rustls/src/conn.rs

```diff
@@ -894,23 +872,20 @@ impl CommonState {
     /// Process `msg`.  First, we get the current state.  Then we ask what messages
     /// that state expects, enforced via `check_message`.  Finally, we ask the handler
     /// to handle the message.
-    fn process_main_protocol<Data>(
+    fn process_main_protocol<Data: SideData>(
         &mut self,
         msg: Message,
         mut state: Box<dyn State<Data>>,
         data: &mut Data,
     ) -> Result<Box<dyn State<Data>>, Error> {
         // For TLS1.2, outside of the handshake, send rejection alerts for
         // renegotiation requests.  These can occur any time.
-        if self.traffic && !self.is_tls13() {
-            let reject_ty = match self.is_client {
-                true => HandshakeType::HelloRequest,
-                false => HandshakeType::ClientHello,
-            };
-            if msg.is_handshake_type(reject_ty) {
-                self.send_warning_alert(AlertDescription::NoRenegotiation);
-                return Ok(state);
-            }
+        if self.traffic
```

Comment:
@djc, I looked at this in more detail and I think there's a more general issue that, if resolved, would make this moot. Simply, this code shouldn't be here.

First of all, we're probably TLS 1.3, and we shouldn't have this TLS-1.2-specific code in the being executed, even just the "self.traffic && !is_tls13()" part, in this common, somewhat hot, code path.  In general, we should be getting rid of any "if this is TLS 1.2 { ... }` blocks in the common path, both to make the correctness analysis easier, and for whatever performance improvement and code bloat elimination we can eek out.

I suggest we move these checks to the `impl State` implementations. I believe if we do that, then the "is client" or "is server" determination is determined by the context (client/tls12.rs or server/tls12.rs) in which the check is being done.

Also, I wonder if either the old or new code is correct. The TLS 1.2 spec says this about the client receiving HelloRequest:

>      The HelloRequest message MAY be sent by the server at any time.

and

>  This message will be ignored by the client if the client is currently negotiating a session.  This message MAY be ignored by the client if it does not wish to renegotiate a session, or the client may, if it wishes, respond with a no_renegotiation alert.

What happens if we're a TLS 1.2 client and the server sends a `HelloRequest` during the handshake? It seems like we're supposed to ignore it and not send an alert. It doesn't look like that's what we do, as I searched the code for `HelloRequest` and didn't see any "ignore the HelloRequest if we're handshaking" logic. Maybe it's intentional that we don't ignore renegotiation requests that are made before the handshake finishes, but if so we should document that. Otherwise, if we are going to implement "ignore the HelloRequest if we're handshaking" logic then we could insert the "send the warning alert if the handshake is done" aspect at that point. I hope that would be out of the commonly-executed code path.

Reply:
Strawman: Continue to reject renegotiation requests that occur during the handshake, and then put this logic in the `ExpectTraffic` implementations in each of client/tls12.rs and server/tls12.rs.

Reply:
See PR #943.